### PR TITLE
[reg, top] cdc handling directly in regfile

### PR DESCRIPTION
**Pull request by tjaychen**
Implement CDC handling directly in regfile.

Reference this [proposal presentation](https://docs.google.com/presentation/d/1o-3Kf7DkppXVgwrkHhGP4HnaXTx2emTgb3XxXXgVAOk/edit?usp=sharing)

### CODE REVIEW

File: util/reggen/bus_interfaces.py
Comment (rswarbrick) [inline util/reggen/bus_interfaces.py:17]: Rather than making these dictionaries map to `Optional[str]`, can we make them `Dict[Optional[str], str]`, but not add an entry if the interface doesn't have an async clock? I think that might be a bit clearer. We'd sample with something like
```python
    async_clk = self.device_async.get(bus_name)
    if async_clk is not None:
        # Do something with this clock
    else:
        # Fully synchronous
```

File: util/reggen/clocking.py
Comment (rswarbrick) [inline util/reggen/clocking.py:43]: What does the `clock` argument do here?

File: util/reggen/clocking.py
Comment (rswarbrick) [inline util/reggen/clocking.py:65]: I think we expect all clocks to match `clk_.*_i`. Can we move that check to the `ClockingItem` constructor (raising an error on failure), and turn this into a simple getter?

File: util/reggen/clocking.py
Comment (rswarbrick) [inline util/reggen/clocking.py:57]: Maybe this should be something like `get_clock_base_name`? A `ClockingItem` might just contain a reset, so I think the method name needs to be a bit more specific.

File: util/reggen/clocking.py
Comment (rswarbrick) [inline util/reggen/clocking.py:59]: Maybe something like "Cannot extract clock base name: this clocking item has no clock" ? The current text looks a bit like we're calling a method a "None clock", which isn't quite what's going on.

File: util/reggen/register.py
Comment (rswarbrick) [inline util/reggen/register.py:448]: I think this returns `Optional[ClockingItem]`?

File: util/reggen/multi_register.py
Comment (rswarbrick) [inline util/reggen/multi_register.py:150]: Like the `Register` code, I think this returns `Optional[ClockingItem]`?

File: util/reggen/multi_register.py
Comment (rswarbrick) [inline util/reggen/multi_register.py:154]: I agree with the comment. Maybe it would be worth adding an assertion to check it's true?

File: util/reggen/reg_block.py
Comment (rswarbrick) [inline util/reggen/reg_block.py:63]: Nit: "block"

File: util/reggen/reg_block.py
Comment (rswarbrick) [inline util/reggen/reg_block.py:88]: Rather than calling a "handle_async" function separately, could this code be folded into `add_raw`? It avoids having to call two methods in order to perform one action (add some registers to the block).

File: util/reggen/reg_block.py
Comment (rswarbrick) [inline util/reggen/reg_block.py:166]: I think this check could be moved into `add_raw`, where we can check the register being added? That gets rid of the validate method that needs calling afterwards, which is a bit nicer. I think it also makes the code a bit simpler because we don't have to iterate when checking.

File: util/reggen/register.py
Comment (rswarbrick) [inline util/reggen/register.py:108]: Could we pass the clocking information to the Register constructor so that we can do the lookup here and get rid of the `handle_async` function? That way, we avoid having a partially constructed object that needs "finishing off" later.

**Review by rswarbrick — COMMENTED** [review]
Some comments on the Python, but this looks like a good plan to me.

**Comment by tjaychen** [comment]
> Some comments on the Python, but this looks like a good plan to me.

thanks for all the responses @rswarbrick. let me look this over again, quite honestly i barely remember what i did.

File: util/reggen/clocking.py
Comment (tjaychen) [reply util/reggen/clocking.py:43]: i feel like this was code left over when i was trying to do something else here...removing it completely.

File: util/reggen/clocking.py
Comment (tjaychen) [reply util/reggen/clocking.py:65]: good idea. let me do that.

File: util/reggen/clocking.py
Comment (tjaychen) [reply util/reggen/clocking.py:57]: yep good point. although combined with the thing above, i should just reference the variable directly.

File: util/reggen/multi_register.py
Comment (tjaychen) [reply util/reggen/multi_register.py:150]: so dummy question here, what in general are the rules for when to explicitly call out the object name vs using `object`?
i ask only because lint didn't seem to flag this, and I'm wondering if it's supposed to do.

File: util/reggen/register.py
Comment (tjaychen) [reply util/reggen/register.py:448]: ah i originally had that, but i got this during lint 

```register.py:460: error: Incompatible return value type (got "object", expected "Optional[ClockingItem]")```

File: util/reggen/bus_interfaces.py
Comment (tjaychen) [reply util/reggen/bus_interfaces.py:17]: got it.

File: util/reggen/multi_register.py
Comment (tjaychen) [reply util/reggen/multi_register.py:154]: good idea.

File: util/reggen/reg_block.py
Comment (tjaychen) [reply util/reggen/reg_block.py:63]: fixed.

File: util/reggen/register.py
Comment (tjaychen) [reply util/reggen/register.py:108]: yep! this one and the two above i'm doing together. It actually does remove lots of code.

File: util/reggen/reg_block.py
Comment (tjaychen) [inline util/reggen/reg_block.py:221]: @rswarbrick this part is kind of annoying.
I originally followed some examples online to pickout variable number of arguments to supply. 

However, `mypy` just kept complaining about function and callable typing issues. In the end I gave up and just made it optional on everyone.

File: util/reggen/multi_register.py
Comment (rswarbrick) [reply util/reggen/multi_register.py]: Well, `object` is a super-type that allows any value at all (so doesn't really provide any information). You can always treat a value as type `object`, but the problem happens when you try to use the result of the function. If we ran MyPy on the call-site (which we don't, because it's in topgen, I think), we'd see it complain about trying to do "ClockingItem stuff" with something whose type it doesn't know.

My general approach is to use `object` when we genuinely don't know the type (for example, we've just read something from an hjson/yaml file) and refine to a useful type with checks (and error messages!). If MyPy isn't clever enough to follow some logic about why "`x` **must** be of type `T` because the code just did XYZ", I insert some assertions.

File: util/reggen/register.py
Comment (rswarbrick) [reply util/reggen/register.py]: Yep, that's because `Clocking.get_by_clock` was returning an `object` rather than an `Optional[ClockingItem]`.

File: hw/ip/prim/rtl/prim_subreg_cdc.sv
Comment (tomroberts-lowrisc) [inline hw/ip/prim/rtl/prim_subreg_cdc.sv:37]: We might want a stipulation / checking on the relationship between src/dst resets where this is used since a dst reset without a src reset would probably lock up in a busy state. That shouldn't be an issue with the intended aon domain targets but could be if we were syncing into some other minor domain which could be reset independently of the bus.

File: hw/ip/aon_timer/rtl/aon_timer.sv
Comment (tomroberts-lowrisc) [inline hw/ip/aon_timer/rtl/aon_timer.sv:73]: I think this comment is out of date

File: hw/ip/aon_timer/rtl/aon_timer.sv
Comment (tomroberts-lowrisc) [inline hw/ip/aon_timer/rtl/aon_timer.sv:131]: Comment no longer required

**Review by tomroberts-lowrisc — APPROVED** [review]
I have only reviewed the aon_timer parts in detail, but they LGTM

**Comment by rswarbrick** [comment]
Hi Tim! Thanks for the changes. I've just left notes about a couple of other things I noticed. Also, there's still the pending question about simplifying async validation (https://github.com/lowRISC/opentitan/pull/7331#discussion_r678145418) from the first review, which I think might be worth doing. (It should actually simplify the code, I think).

File: util/reggen/reg_block.py
Comment (rswarbrick) [reply util/reggen/reg_block.py:221]: Well, the point is that `_handle_reserved` is put in a dictionary of "handlers", which are all called with the same arguments. So MyPy is actually complaining correctly: if we removed the `clocks` argument then the call at the end of `add_raw` would fail at runtime. I don't think the private method is called in any other context, so I'd suggest just making the type signature match the other handlers.

Comment (tjaychen) [reply util/reggen/reg_block.py:221]: i should have clarified i was doing [this](https://stackoverflow.com/questions/50496821/how-to-call-functions-in-a-dict-that-take-different-numbers-of-arguments).  And it actually worked just fine, but mypy just didn't like the typing. 

The main thing it was complaining about (which i could not figure out) is that the input argument was a function and not a callable.  That was...odd to me, because I thought all functions were callables.

**Comment by tjaychen** [comment]
> Hi Tim! Thanks for the changes. I've just left notes about a couple of other things I noticed. Also, there's still the pending question about simplifying async validation ([#7331 (comment)](https://github.com/lowRISC/opentitan/pull/7331#discussion_r678145418)) from the first review, which I think might be worth doing. (It should actually simplify the code, I think).

o i actually simplified the async validation already. It is no longer called as a separate step, instead it is called now as part of the `_handle_register` process. So there's no extra iteration anymore. I just kept it as a separate function because there was a giant error message and i wanted to keep it separate for readability.

File: hw/ip/aon_timer/rtl/aon_timer.sv
Comment (tjaychen) [reply hw/ip/aon_timer/rtl/aon_timer.sv:186]: o yeah i think you're right.  let me change this.

File: hw/ip/prim/rtl/prim_subreg_cdc.sv
Comment (tjaychen) [reply hw/ip/prim/rtl/prim_subreg_cdc.sv:37]: yeah good point, i'm adding an assertion for the busy.

File: hw/ip/aon_timer/rtl/aon_timer.sv
Comment (tjaychen) [reply hw/ip/aon_timer/rtl/aon_timer.sv]: fixed!

File: hw/ip/aon_timer/rtl/aon_timer.sv
Comment (tjaychen) [reply hw/ip/aon_timer/rtl/aon_timer.sv]: fixed!

File: util/reggen/bus_interfaces.py
Comment (tjaychen) [reply util/reggen/bus_interfaces.py:80]: ah got it. will do.

File: util/reggen/clocking.py
Comment (tjaychen) [reply util/reggen/clocking.py:41]: fixed.

File: hw/ip/prim/rtl/prim_subreg_cdc.sv
Comment (msfschaffner) [inline hw/ip/prim/rtl/prim_subreg_cdc.sv:76]: quick question, am I understanding correctly that the `src_busy_q && src_ack` condition is only present to make sure we're sampling the dst side as soon as possible in order to not loose another dst cycle? i.e. `src_ack` acts like a `src_update_i` indication?

File: hw/ip/prim/rtl/prim_subreg_ext_async.sv
Comment (msfschaffner) [inline hw/ip/prim/rtl/prim_subreg_ext_async.sv:41]: nit: desitnation -> destination

File: hw/ip/prim/rtl/prim_subreg_cdc.sv
Comment (tjaychen) [reply hw/ip/prim/rtl/prim_subreg_cdc.sv:76]: yep! this is basically to cover the case that after we ack on a software initiated transaction, they can immediately read back what they wrote.

Comment (msfschaffner) [reply hw/ip/prim/rtl/prim_subreg_cdc.sv:76]: ok, got it, LGTM!

**Review by msfschaffner — APPROVED** [review]

File: hw/ip/prim/rtl/prim_subreg_cdc.sv
Comment (eunchan) [inline hw/ip/prim/rtl/prim_subreg_cdc.sv:9]: CDC implementation looks good to me.

Thanks Tim for implementing this feature!

**Review by eunchan — APPROVED** [review]

File: util/reggen/uvm_reg_base.sv.tpl
Comment (tjaychen) [inline util/reggen/uvm_reg_base.sv.tpl:492]: @cindychip @weicaiyang do you mind having a look at this?
basically the async registers now have a wrapper level on top of `prim_subreg`, so i'm guessing this is what we need to do.

File: util/reggen/reg_block.py
Comment (rswarbrick) [reply util/reggen/reg_block.py:224]: I suspect that it was complaining that it wasn't a callable with the expected arguments. The point is that MyPy types `def foo(a: int) -> object` as `Callable[[int], object]` and types `def bar(a: int, b: str) -> object` as `Callable[[int, str], object]` and these are different types.

You **could** do something like add an explicit type to the definition of the list, which might work (`Union[Callable[[int], object], Callable[[int, str], object]]`), but it's probably a sign that we're doing something unnecessarily complicated.

File: util/reggen/uvm_reg_base.sv.tpl
Comment (weicaiyang) [reply util/reggen/uvm_reg_base.sv.tpl:492]: This LGTM, thanks @tjaychen
If this path is wrong, the `csr_hw_reset` test will fail as it checks the hierarchy path.

**Comment by vogelpi** [comment]
Sorry guys, I did not manage to review this yet but definitely would like to have a closer look tomorrow. But I don't want to block merging this. I can also provide feedback post merge if you want to go ahead now.

**Comment by tjaychen** [comment]
thanks everyone.
@vogelpi , i'm going to merge this for now, but please feel free to have a second look, and we can continue adjusting for any concerns that show up.

File: hw/ip/prim/rtl/prim_subreg_cdc.sv
Comment (vogelpi) [inline hw/ip/prim/rtl/prim_subreg_cdc.sv:103]: This input is just used to gate an SVA checking that `req_i` never goes low without seeing ACK. This should never happen as it can have very undesirable side effects such as leaving the primitive in an undetermined state, data being lost in the CDC etc. If this ever happens, it indicates a bug in the logic driving the `req_i`. So far, there has been only one scenario in which this check should be disabled and that is during an LC escalation, i.e., when we don't care about the CDC anymore (see #7333). I think driving a constant 0 here isn't a good idea.

What probably should be done is to add a new input to the auto-generated `*reg_top.sv` to gate this SVA only during  an LC escalation but that will mean quite some work. If there is another way for us to guarantee that the auto-gen regfile never lowers this REQ without ACK we are fine. But for the `prim_subreg_ext_async` we probably need it as IP designers will connect that to custom logic that is not auto-generated/not under control of the regtool.

Comment (vogelpi) [reply hw/ip/prim/rtl/prim_subreg_cdc.sv:103]: Ah, now I see that you have added an SVA for `src_busy_o` below to cover this case. This looks good but I suspect that this SVA also needs to be gated during LC escalation. Maybe @cindychip can comment on that.

**Review by vogelpi — APPROVED** [review]
Thanks @tjaychen for working on this and for the useful comments in the code!. The PR looks good to me.

I have one comment regarding the SVA ensuring that `req_i`/`src_busy_o` stays high until we see ACK. It's nothing fundamental, but it may need a follow-up PR to disable it during an LC escalation only.

File: hw/ip/prim/rtl/prim_subreg_cdc.sv
Comment (tjaychen) [reply hw/ip/prim/rtl/prim_subreg_cdc.sv:103]: ah i should have clarified here. So originally, the `prim_sync_reqack` was used with `EnStableData` set to 0. 
The only reason for this... is because the `src_req` is really just a write pulse from the regfile. 

So the behavior I really wanted is a cross between `prim_pulse_sync` and `prim_reqack`.  I suppose I could have just added a flop in front honestly to reflect that behavior... but yeah that's how we ended up here.

Yeah good point on LC escalation disable, I think we might need to do that especially if LC escalation causes the whole system to reset.  Actually in general, how do we want to handle that case?

**Pull request merged**
